### PR TITLE
Make an asynchronous OpenCL code

### PR DESCRIPTION
**Pull request by SukharevPavel**
Pull request for issue: https://github.com/ethereum-mining/ethminer/issues/1366

Making an asynchronous OpenCL work.
If the solution is found, or the new work is received, the kernel stops as fast as it can.
The new work is received in a separate thread, thus solving the problem with blocking enqueueReadBuffer call. 
It helps to eliminate stale work processing, and ensure fast share submitting. Also, this approach allows to use a large worksizes. 

I don't know GCN assembler, so I can't implement kernel changes for binary kernels. But they work fine with this code (although, no speedup).

Hash are counted on a kernel side, I'm using result array to store hash count. The problem is that we should reinitialize result array every mining cycle to have actual hash speed, which could impact the performance, especially for small work sizes (but I didn't notice it). I think, we could implement hash reading not on every cycle, thus decrease that overhead. For example, if we would read hash count every 1 second, it would be enough to make that overhead insignificant.
 
The miner works fine for both kernel types, but for -Z option it gets some wrong GPU results on low difficulties. I guess, it's because of asynchronous work receiving, which is instant in case of -Z mode, and we can do nothing about it.

### CODE REVIEW

**Comment by SukharevPavel** [comment]
@jean-m-cyr Can you help me with binary kernels?

**Comment by jean-m-cyr** [comment]
@SukharevPavel Not at the moment (busy with current implementation). Maybe later. I need to take the time to better understand the pro's and con's of your approach.

**Comment by SukharevPavel** [comment]
@jean-m-cyr I understand. I'll try my best to implement the binary kernel changes by myself for now.

**Comment by jean-m-cyr** [comment]
@SukharevPavel 

> I don't know GCN assembler, so I can't implement kernel changes for binary kernels. But they work fine with this code (although, no speedup).

Why would you make the effort if there is no speedup?

I see what you're trying to do. But the way I see it, without a synchronous loop you will always have an issue with host CPU usage because of the need to poll for a result.

**Comment by SukharevPavel** [comment]
@jean-m-cur
I mean, there is no speedup for binary kernel right now. 

>you will always have an issue with host CPU usage because of the need to poll for a result

I dont get it. Where exactly in the code, as you think, CPU will be wasted?
See ethash.cl. It automatically performs a fast finish when result[MAX_RESULTS] != 0, what lead to faster OpenCL cycle and thus faster enqueueReadBuffer read. As I mentioned before, my experiment shows that work-item ends work 1400 times faster when this case is fulfilled.
Result[MAX_RESULTS] gets a non-null value:

1) when GPU found a new result and write it to buffer result

In this case, no CPU power is wasted obviously, there is only GPU work and blocking CPU read.

2) when miner receives a new work

Here I'using kick_miner() function, which fires only  once when new work came. This function was in a miner for a long time, but it was not implemented in CLMiner.cpp for some reason. We changing result buffer with enqueueWriteBuffer from another OpenCL queue, which is not busy and can execute the query as fast as it can (m_invalidatingQueue). That sets result[MAX_RESULTS] to 0xffffffff, and, according to the aforementioned algorithm, forces the kernel to end fast. We don't spend any excessive CPU power here either.

I check CPU usage both for default ethminer from master branch and this code with ‘ps‘ command. And new miner don't show any excessive CPU usage.

**Comment by jean-m-cyr** [comment]
Quick comment. iterations is always set to 1. This GPU code will not exit an faster than before.

**Comment by SukharevPavel** [comment]
>quick comment
It's 7 am here for now :)

I didn't check the new kernel code logic much, but isn't iterations are about counting different nonces on one work-item? If it is so, then we had only one iteration on the old kernel too. The speedup is achieved because we don't access dag and make speed counting, we just fast end the kernel.

**Comment by jean-m-cyr** [comment]
You are not exiting the GPU code any faster. max iterations is always = 1 in Ethminer. The loop will only do one pass regardless of the state of the terminate flag!

**Comment by SukharevPavel** [comment]
@jean-m-cyr 
The terminate flag is before the loop. It checks the global variable.
Ok, let's look at example. We have 10000 work-groups, and we found a solution for work-item in 110th work-group. This changes g_output[MAX_OUTPUTS] to 1, right? After that every single work-item in a work-group which will tart to execute on the device after that will instantly meet that condition and do a fast finish. So, we got 110+x work-groups which made a full work (x is for number of work-groups which were already executed on device when solution was found), and 10000-x-110 groups which will finish instantly.
In the OpenCL, if number of work-items is larger than device capacity (for example, this is typical for AMD GPU to have 36 CU, which had 4 vector units which execute wavefronts of 64 items), the work-groups which don't receive processing their time, will wait for execution until GPU finish computations of current work-items and free some place for them.
And my experiments showed so. In the issue I attached a screen of miner's work, when cycles are finished in 7 seconds in case no interruption, and much faster if the solution found or new work is received.

**Comment by SukharevPavel** [comment]
@jean-m-cyr 
Ok, I've just found a mistake in PR. I've always used blocking `m_invalidationQueue.enqueueWriteBuffer` call, but before the PR I decided to change it to non-blocking. I decided to do that because I didn't want to interrupt main thread. But this breaks the behavior I described above. So, we should return blocking call there. In that case, the mining process looks like I told earlier: https://ibb.co/bVJ9KT

**Comment by jean-m-cyr** [comment]
Ah, ok I see. You are not trying to terminate an active work group early. You are only terminating all pending work groups early. That has advantages, but here's something to think about. In order for the gpu to run 7 seconds without terminating you'd launch a huge number of work groups. That's fine, but each work group now has to check the termination flag which lives in global memory and requires a PCIe transfer from host memory. Now suppose you have 16 or more gpus on your rig and you need to terminate early after running only 100 ms. It would result in a huge number of uncached PCIe reads all at once. Would this be an issue?

**Comment by SukharevPavel** [comment]
@jean-m-cyr 
I tested the performance difference between default kernel without if-case and modified kernel. I launched -G -M with warmup time 60 and testing time 600 (20 rounds, each takes 30 seconds). I got:

OpenCL kernel | Max speed, H/s | Mean speed, H/s
-- | -- | --
Common | 27069502 | 26841241
Modified | 27069502 | 26774476

I'm almost sure that existing difference can be explained by calculation heterogeneity. For example, in another test, I got performance advantage for modified kernel, but I forgot to write the exact results. Also, I did that tests only for the old kernel code. I guess, I should take another test for the current kernel code.

**Comment by SukharevPavel** [comment]
@jean-m-cyr 
Oh sorry, I forgot the second part. As I could see, the result buffer are allocated for each device separately, and accessed separately, so there is should be no problem for big GPU rigs. But I never worked with more than 1 GPU, so I can be wrong.

**Comment by SukharevPavel** [comment]
I've just make a test for miner from master branch and for modified kernel from PR. The command is:
./ethminer -G -M 1 --cl-nobinary --benchmark-warmup 30 --benchmark-trial 50

The results are described below:

OpenCL kernel | Max speed, H/s | Mean speed, H/s
-- | -- | --
Common | 28325376 | 28016371
Modified | 28325376 | 28142262

So, as I told earlier, the speed change can be explained mostly by the fact that calculation process is not smooth by itself.

**Comment by SukharevPavel** [comment]
>If a GPU thread finds a solution it must terminate all the workgroups before the read will unblock. No?

Yes. We have the condifion in the OpenCL code
```
if (g_output[MAX_OUTPUTS]!=0) {
return;
}
```
So, when the solution is found, g_output[MAX_OUTPUTS] increments to 1 (according to normal kernel logic), and it will end all awaiting workgroups as soon as they get processing time.

**Comment by jean-m-cyr** [comment]
So, suppose you launch 500,000 works groups so they will run a long time. The 1st workgroup then finds one solution. The remaining 499,996 (supposing you run 4 waves) will almost instantly read g_output[MAX_OUTPUTS] from global memory. That's 499,996 data transfers via the PCIe bus.

**Comment by SukharevPavel** [comment]
@jean-m-cyr 
Yes, still it would much faster than processing the full ethash algo. Do you need some kind of time comparison? I've already showed the timestamps for big and small works. Compare this 499.996 data transfers with 499.996 multiply by the 64 accesses to DAG at its random positions, plus some hash calculations. And if the new work is received by the miner, these calculations are completely irrelevant, because they used stale header value.

And the problem should not appear with increasing GPU number, cause, if I understand it right, the __global OpenCL modifier means it is global for the whole GPU, not he whole system itself. So, the 16 GPU will make their 499996 calls separately inside themselves.

**Comment by SukharevPavel** [comment]
I found one more error in my PR. Since kick_miner() is called is a separate thread, which is common for all miner instances, we should move the init() logic out of there, otherwise it will block the thread and provide sequential DAG initialization behavior for GPU rig as a result.

**Comment by SukharevPavel** [comment]
Fix the host-part bugs that I've found and described above.

**Comment by jean-m-cyr** [comment]
Ok, I've had a chance to play with this and have some questions and suggestions:

- You are doing one 4 byte PCIe read per hash to check the abort flag, and one 4 byte atomic increment (one read + one write = 8 bytes) per hash. That's a total of 12 bytes transferred over PCIe per hash. On a 450 MH/S rig, that's 12 x 450 MB/s = 5.4 GB/S transferred over PCIe. Well below the PCIe 1.1 250GB/S data rate. So that should not be a problem.

- However the atomic increment serializes the GPU threads, temporarily preventing them from running in parallel and slowing things down. I'd suggest only incrementing the hash counter once per wave (get_local_id() == 0). That would result in 64 times less contention for the atomic increment and would reduce the amount of PCIe IO drastically.

- question: Why did you move all of the init stuff out of the workloop and into the kick_miner method. Wouldn't it have been simpler to just write the abort flag synchronously to the abort command queue and leave the workloop as is? Also why create a new thread for the kick_miner processing? A synchronous write to an empty queue will take in the order of microeconds to return so you don't need to worry about blocking the calling host thread.

**Comment by SukharevPavel** [comment]
>I'd suggest only incrementing the hash counter once per wave (get_local_id() == 0)

Yes, I thought about it. Since I didn't see any speed reduction with simple atomic_inc, I decided to leave it as it is, just for simplicity, cause I couldn't measure any difference. So, we can do it to reduce memory transfers.

>Wouldn't it have been simpler to just write the abort flag synchronously to the abort command queue and leave the work loop as is?

In the current miner version we wouldn't update the work in this case. Look, in the current miner we have an order:
1. enqueueNDRangeKernel
2. check for work
3. enqueueReadBuffer (blocking)

Let's suppose that we place abort command in kick_miner. In this case, we drop the enqueueReadBuffer call when the new work came, but we don't actually change the work before the new enqueueReadBuffer call, so it will be launched with old parameters. Of course, we can just move checking for work after enqeueReadBuffer and before enqueueNDRangeKernel, but in this case we would lost a couple of milliseconds to work switching.

I guess, we can do something with it, but since I don't understand all the thin logic about work changes (for example, I have never got the epoch change during my experiments, so I was unsure to split init() call from another logic for new block), I decided to move the whole work to kick_miner. Of course, it was just in sake of preserving the stable logic, I haven't any special ideas about it. 

> Also why create a new thread for the kick_miner processing? 

I thought that blocking one of the main threads is just not right. And, in case of init() call in kick_miner body, this is justified. 
The problem is, that, for example, if we have 32 GPUs, then the 32th GPU will get its abort command only after 31 another GPUs. I think, it can lead to work terminating delay for about 20 ms for that GPU. But I didn't measure how much will it cost to create a new thread, so I don't know if it is right.

>Why run with such a large multiplier. Instead, why not run with a small multiplier but run each work group in an infinite loop (iterations = infinity)

Yes, that's a great idea, cause it will make OpenCL kernel finish it's work almost instantly in case of new block, and we will never lose any time for making cyclic enqueueNDRangeKernel calls. I didn't think about it, cause I was mostly worried about my solution viability, so I wanted to make it as simple as possible without many code changes.

**Comment by jean-m-cyr** [comment]
DAG is in GPU memory. The g_output pointer points to shared memory, which resides on the host and is accessed by the GPU via a PCIe mapped window into host memory. Notice that we don't create a cl::buffer for the DAG, but we do for the result buffer.

**Comment by jean-m-cyr** [comment]
You seem to be correct, I thought the CL_MEM_USE_HOST_PTR was used on the cl::buffer instantiations. So yes the response buffer will reside in GPU memory. PCIe transfers would only occur when the host accesses the buffer.

**Comment by SukharevPavel** [comment]
@jean-m-cyr 
I have encountered the problem while trying to implement infinite search loop for ethash algorithm.
In infinite loop we should update working nonce with changing gid variable for every cycle. The kernel version with `iterations` variable had used:
```
const uint gid = get_global_id(0) + iter * get_num_groups(0) * WORKSIZE;
nonce = start_nonce + gid
```
for that, but we shouldn't do it, because the most of the groups will never get any processing time. For example, AMD RX 480 can process only 27648 work-items simultaneously, and these work-items will take the GPU processing elements forever until the break. So, for local_work = 192 and global_work we will get:
item with get_global_id(0) = 0 will process nonces: `start_nonce, start_nonce + 192*16384, start_nonce + 2*192*16384`
item with get_global_id(0) = 1 will process nonces: `start_nonce + 1, start_nonce + 1 + 192*16384, start_nonce + 2*192*16384 + 1`
...
item with get_global_id(0) = 27647 will process nonces: `start_nonce + 27647, start_nonce + 27647 + 192*16384, start_nonce + 2x192*16384 + 27647`

So, basically, we will skip a lot of nonce values (for example, range `[start_nonce+27648, start_nonce + 192*16384]` will never be processed), which is not good. Actually, we always skip range `[start_nonce+27648 + 192*16384 * x, start_nonce + (x+1) * 192 * 16384]` where `x is every positive integer`

I have two solutions to this problem:
1) Have global 64-bit variable add access it through atomic_inc when we want a new value for int . This solution is obviously bad.

2) Determine optimal global_work_size on the fly. If we set global work size to 27648 (which is 144 groups with size 192), this problem will not be exist for AMD RX 480. Sadly, I don't know how to determine the actual number of processing elements for different devices. Do you have any idea about it?

**Comment by jean-m-cyr** [comment]
It doesn't matter if we skip nonces... All nonces are equally likely to produce a solution. As long as we search an equal number of nonces, the probability of finding a good one stays the same.

**Comment by SukharevPavel** [comment]
Since https://github.com/ethereum-mining/ethminer/pull/1400 was merged, this PR can be closed.

**Comment by jean-m-cyr** [comment]
Closed as requested by author
